staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go

```diff
@@ -18,84 +18,70 @@ package configuration
 
 import (
 	"fmt"
-	"reflect"
 	"sort"
-
-	"github.com/golang/glog"
+	"sync/atomic"
 
 	"k8s.io/api/admissionregistration/v1beta1"
-	"k8s.io/apimachinery/pkg/api/errors"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	admissionregistrationinformers "k8s.io/client-go/informers/admissionregistration/v1beta1"
+	admissionregistrationlisters "k8s.io/client-go/listers/admissionregistration/v1beta1"
+	"k8s.io/client-go/tools/cache"
 )
 
-type MutatingWebhookConfigurationLister interface {
-	List(opts metav1.ListOptions) (*v1beta1.MutatingWebhookConfigurationList, error)
-}
-
 // MutatingWebhookConfigurationManager collects the mutating webhook objects so that they can be called.
 type MutatingWebhookConfigurationManager struct {
-	*poller
+	configuration *atomic.Value
+	lister        admissionregistrationlisters.MutatingWebhookConfigurationLister
 }
 
-func NewMutatingWebhookConfigurationManager(c MutatingWebhookConfigurationLister) *MutatingWebhookConfigurationManager {
-	getFn := func() (runtime.Object, error) {
-		list, err := c.List(metav1.ListOptions{})
-		if err != nil {
-			if errors.IsNotFound(err) || errors.IsForbidden(err) {
-				glog.V(5).Infof("MutatingWebhookConfiguration are disabled due to an error: %v", err)
-				return nil, ErrDisabled
-			}
-			return nil, err
-		}
-		return mergeMutatingWebhookConfigurations(list), nil
+func NewMutatingWebhookConfigurationManager(informer admissionregistrationinformers.MutatingWebhookConfigurationInformer) *MutatingWebhookConfigurationManager {
+	manager := &MutatingWebhookConfigurationManager{
+		configuration: &atomic.Value{},
+		lister:        informer.Lister(),
 	}
 
-	return &MutatingWebhookConfigurationManager{
-		newPoller(getFn),
-	}
+	// Start with an empty list
+	manager.configuration.Store(&v1beta1.MutatingWebhookConfiguration{})
+
+	// On any change, rebuild the config
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(_ interface{}) { manager.updateConfiguration() },
+		UpdateFunc: func(_, _ interface{}) { manager.updateConfiguration() },
+		DeleteFunc: func(_ interface{}) { manager.updateConfiguration() },
+	})
+
+	return manager
 }
 
 // Webhooks returns the merged MutatingWebhookConfiguration.
-func (im *MutatingWebhookConfigurationManager) Webhooks() (*v1beta1.MutatingWebhookConfiguration, error) {
-	configuration, err := im.poller.configuration()
-	if err != nil {
-		return nil, err
-	}
-	mutatingWebhookConfiguration, ok := configuration.(*v1beta1.MutatingWebhookConfiguration)
-	if !ok {
-		return nil, fmt.Errorf("expected type %v, got type %v", reflect.TypeOf(mutatingWebhookConfiguration), reflect.TypeOf(configuration))
-	}
-	return mutatingWebhookConfiguration, nil
+func (m *MutatingWebhookConfigurationManager) Webhooks() *v1beta1.MutatingWebhookConfiguration {
```

Comment:
If the informer encounters errors, we should surface them here. We don't want an apiserver to apply stale webhook configurations if the informer has been out of sync for a long time. I don't know how feasible it is to expose informer errors though. @liggitt wdyt?

I checked the rbac authorizer code, it didn't surface any informer error. It's ok for the rbac authorizer, since it's using a loopback client so the chance of informer being stale is low. But the webhook configuration manager might run on a different machine than the kube-apiserver, so we need to handle informer errors.

Reply:
This way is much like controllers usage.  They all wait for cache sync before use, and then the add/update/delete operations handled by event handler. Your meaning is more like a network issue, which is handled by lower level, right?

Reply:
Kubernetes controllers only need to guarantee eventual consistency. Controllers can retry after the network issue is resolved. But for the webhook admission controller, if it makes admission decision based on stale configuration, it doesn't have a chance to retry.

On the other hand, maybe we can argue that if an network issue affects the informer, then aggregator won't be able to delegate the API request to the extension apiserver either.

Reply:
The PR lgtm except this issue. I'll be OOO until the new year, so I'm re-assigning it to @cheftako.

Reply:
> If the informer encounters errors, we should surface them here. We don't want an apiserver to apply stale webhook configurations if the informer has been out of sync for a long time. 

@sttts @hzxuzhonghu Have we started a solution for this issue? I can start a design on how to expose apiserver's observed configurations. We mentioned CRD needed a similar mechanism in the sig meeting, what are the exact requirements of CRD?

Reply:
Not yet.

Reply:
We're only talking about degrees of staleness.  Without server fencing or checkpointing of some kind, you face the same problem.  I think the rbac comparison is very pertinent since the same arguments about "you shouldn't have been able to create this" apply.  When an aggregated apiserver loses its connection to the kube-apiserver, it's going to start rejecting requests on 403's anyway.

While I think the idea of exposing the state of the lister is good idea, I don't think it should block the overall improvement we get out of this pull.